File: registration-form-dotnet/server2/Services/UserServices/Adapters/UserService.cs
using Microsoft.EntityFrameworkCore;
using server2.Services.UserServices.Data;
using server2.Services.UserServices.Model;
using server2.Services.UserServices.Ports;

namespace server2.Services.UserServices.Adapters
{
    public class UserService : IUserService
    {
        private readonly UserContext userContext;

        public UserService (UserContext userContext)
        {
            this.userContext = userContext;
        }

        public async Task < User? > Login (string username, string password)
        {
            try
            {
                User? userToLogin =
                    await userContext.Users.FirstOrDefaultAsync (
                        u => u.Username == username && u.Password == password);

                if (userToLogin == null)
                {
                    return null;
                }

                // password hash verifications

                return userToLogin;
            }
            catch (Exception e)
            {
                throw new Exception ($"{ e.Message }");
            }
        }

        public async Task < IEnumerable < User >> GetUsers ()
        {
            return await userContext.Users.Select (
                u => new User {
                    Id = u.Id, Username = u.Username, Email = u.Email
                    }).ToListAsync ();
        }

        public async Task < User? > GetUser (int userId)
        {
            try
            {
                User? userToGet = await userContext.Users.FindAsync (userId);
                
                if (userToGet == null)
                {
                    return null;
                }
                
                return userToGet;
            }
            catch (Exception e)
            {
                throw new Exception ($"{ e.Message }");
            }
        }

        public async Task < User? > AddUser (User user)
        {
            try
            {
                if (user == null)
                {
                    throw new ArgumentNullException (
                        nameof (user), "User object cannot be null");
                }

                User? userToAdd = userContext.Users.Where (
                    u => u.Email == user.Email).FirstOrDefault ();

                if (userToAdd != null)
                {
                    throw new IntegrityException (
                        $"User with email { user?.Email } is already registered");
                }

                userContext.Users.Add (user);
                await userContext.SaveChangesAsync ();
                
                return user;
            }
            catch (Exception e)
            {
                throw new Exception ($"{ e.Message }");
            }
        }

        public async Task < User? > UpdateUser (int userId, User user)
        {
            try
            {
                User? userToUpdate = await userContext.Users.FindAsync (userId);

                if (userToUpdate == null)
                {
                    return null;
                }

                userContext.Entry (userToUpdate).CurrentValues.SetValues (user);
                await userContext.SaveChangesAsync ();

                return user;
            }
            catch (Exception e)
            {
                throw new Exception ($"{ e.Message }");
            }
        }

        public async Task < User? > DeleteUser (int userId)
        {
            try
            {
                User? userToDelete = await userContext.Users.FindAsync (userId);

                if (userToDelete == null)
                {
                    return null;
                }

                userContext.Users.Remove (userToDelete);
                await userContext.SaveChangesAsync ();

                return userToDelete;
            }
            catch (Exception e)
            {
                throw new Exception ($"{ e.Message }");
            }
        }
    }
}

File: express/server/src/services/userService.js
const User = require('../models/userModel');


const UserService = {

  createUser: async (userData) => {

    console.log('userData', userData);

    try {

      console.log('userData.email', userData.email);

      let user = await User.findOne({
        where: {
          email: userData.email
        }
      });
    
      if (user === null) {
        
        // Process data before saving

        user = await User.create(userData);
        return user;
    
      } else {
      
        console.log(`Email ${ userData.email } is already registered`);
      
      }
  
    } catch (error) {
            
      throw new Error(`Error creating user: ${ error.message }`);
  
    }

  },
  
  getUsers: async () => {

    try {
      
      const users = await User.findAll();
      return users;
  
    } catch (error) {
      
      throw new Error(`Error getting users: ${ error.message }`);
  
    }
    
  },

  getUser: async (id) => {

    try {
      
      const user = await User.findByPk(id);
      return user;

    } catch (error) {
      
      throw new Error(`Error getting user: ${ error.message }`);
    
    }
  
  },

  updateUser: async (id, userData) => {

    try {
      
      const user = await User.findByPk(id);
      await user.update(userData);
      return user;
    
    } catch (error) {

      throw new Error(`Error updating user: ${ error.message }`);
    
    }

  },

  deleteUser: async (id) => {

    try {
      
      const user = await User.findByPk(id);
      await user.destroy();

    } catch (error) {
      
      throw new Error(`Error deleting user: ${ error.message }`);

    }

  }

};

module.exports = UserService;
